### ISSUE TITLE
Clarification on reset when collisions occur

### ISSUE THREAD

**Issue by tommasomarzi** [bassamlab/SigmaRL#1]
Hello,

Thank you for the repository, the environment you proposed is very interesting!

I have a question about what happens when an agent collides. In the paper, you mention that you reset colliding agents with random states, while in the docstring (lines 89-94 of `road_traffic.py`) you wrote that **all** the agents will be reset (not only the colliding ones). 

However, in both cases, from the code, it seems to me that collisions are only used to get the done signals, and they do not affect the observations. Furthermore, in the VMAS function `get_from_scenario()` ([line 130 of `environment.py`](https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/main/vmas/simulator/environment/environment.py#L130)) the `done()` function is called after observations are retrieved.

Am I missing something?

Thanks in advance!

**Comment by Jianye-Xu** [comment]
Hello,

Thanks for your interest in our repo and for raising this question!

Once a collision occurs, we reset all agents during **training** and reset only colliding agents during **testing**. The [docstring of `road_traffic.py`](https://github.com/cas-lab-munich/SigmaRL/blob/ac8e15ea522b313c8448185086353639f4e5c925/scenarios/road_traffic.py#L89) conveys the same meaning (note that "non-testing mode" means "training"). Please double-check.

Regarding the effect of done signals on the observations: The done signals do not affect the observations of the current time step (actually I did not understand why you are asking this question). During testing, I manually call the `reset_world_at` function in my scenario file to reset colliding agents (see [this line](https://github.com/cas-lab-munich/SigmaRL/blob/ac8e15ea522b313c8448185086353639f4e5c925/scenarios/road_traffic.py#L2603)). During training, I let VMAS to handle it by setting the done signals to true. The reason for this is that VMAS resets all agents by default but in my case, I need to pass another argument (`agent_index`) to the `reset_world_at ` function to reset a specific agent during testing. 

Does this answer your questions?

**Comment by tommasomarzi** [comment]
Hi, 

Thanks for the quick reply!

> Once a collision occurs, we reset all agents during training and reset only colliding agents during testing. The [docstring of road_traffic.py](https://github.com/cas-lab-munich/SigmaRL/blob/ac8e15ea522b313c8448185086353639f4e5c925/scenarios/road_traffic.py#L89) conveys the same meaning (note that "non-testing mode" means "training"). Please double-check.

Sorry, I confused the train and test behaviours from the main text. Thank you for pointing this out!

>Regarding the effect of done signals on the observations: The done signals do not affect the observations of the current time step (actually I did not understand why you are asking this question). During testing, I manually call the reset_world_at function in my scenario file to reset colliding agents (see this line). During training, I let VMAS to handle it by setting the done signals to true. The reason for this is that VMAS resets all agents by default but in my case, I need to pass another argument (agent_index) to the reset_world_at  function to reset a specific agent during testing.

I mentioned the done function since in the testing phase it calls reset_world_at, as you also pointed out. As for the training phase, it's not clear to me when VMAS resets the agent positions after collisions: from the `get_from_scenario()` function, it seems to me that next observations are simply returned (without reset) even if done is set to true.

Thanks again for your help!

**Comment by Jianye-Xu** [comment]
> As for the training phase, it's not clear to me when VMAS resets the agent positions after collisions

I slowly understand what is confusing you. I would refer you to the [`rollout`](https://github.com/pytorch/rl/blob/e82a69f5af94cc936c4b872fd2ed499ed33b4f8e/torchrl/envs/common.py#L2315) function and the [`step_and_maybe_reset`](https://github.com/pytorch/rl/blob/e82a69f5af94cc936c4b872fd2ed499ed33b4f8e/torchrl/envs/common.py#L2783) function defined in TorchRL. VMAS relies on them to call the reset function.

Hope that would help you.

**Comment by tommasomarzi** [comment]
Sorry, let me try to explain this better. 

I'm trying to use the road traffic environment using the [use_vmas_env.py](https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/main/vmas/examples/use_vmas_env.py) example. In this case, it seems to me that, when a collision occurs, in training mode there's no reset in the corresponding environment, as observations are generated in the `get_from_scenario()` and subsequently there are no reset calls.

Am I wrong in this?

Thanks for the links, I'll look at them! And thanks again for your patience!

**Comment by Jianye-Xu** [comment]
You understanding that there is not reset calls is correct. I think [`[use_vmas_env.py]`](https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/main/vmas/examples/use_vmas_env.py) is mainly designed for debugging but not for training or testing, and therefore it is not complete. For example, it does not check the done signals and does not call the reset function. I would refer to my repo for an complete example. In my repo, the rollout function is called through:
[main_training.py](https://github.com/cas-lab-munich/SigmaRL/blob/ac8e15ea522b313c8448185086353639f4e5c925/main_training.py#L6) -> [SyncDataCollector](https://github.com/cas-lab-munich/SigmaRL/blob/ac8e15ea522b313c8448185086353639f4e5c925/utilities/mappo_cavs.py#L233) -> [function iterator](https://github.com/pytorch/rl/blob/e82a69f5af94cc936c4b872fd2ed499ed33b4f8e/torchrl/collectors/collectors.py#L960). As said before, the rollout function handles the done signals and calls the reset function.

**Comment by Jianye-Xu** [comment]
Did this answer your question? May I close the issue?

**Comment by tommasomarzi** [comment]
I confirm that my doubt has been solved, since it originated from the script I mentioned earlier. I was using that code because I needed a loop in which the step function was explicitly defined, and having never used TorchRL, I took that as an example. Unfortunately, I haven't had time to check the links you sent me yet, but I'll adapt my code to the TorchRL library.

I mark the issue as solved, thanks again for your availability!
